Solutions/Corvus.Tenancy.Storage.Azure.Blob/Corvus/Tenancy/Storage/Azure/TenantProviderBlobStore.cs

```diff
@@ -116,24 +123,37 @@ public async Task<TenantCollectionResult> GetChildrenAsync(string tenantId, int
         }
 
         /// <inheritdoc/>
-        public async Task<ITenant> UpdateTenantAsync(ITenant tenant)
+        public async Task<ITenant> UpdateTenantAsync(
+            string tenantId,
+            IEnumerable<KeyValuePair<string, object>>? propertiesToSetOrAdd,
+            IEnumerable<string>? propertiesToRemove)
         {
-            if (tenant is null)
+            if (propertiesToSetOrAdd is null && propertiesToRemove is null)
             {
-                throw new ArgumentNullException(nameof(tenant));
+                throw new ArgumentNullException(nameof(propertiesToSetOrAdd), $"{nameof(propertiesToSetOrAdd)} and {nameof(propertiesToRemove)} cannot both be null");
             }
 
-            if (tenant.Id == this.Root.Id)
+            if (tenantId == this.Root.Id)
             {
+                ((RootTenant)this.Root).UpdateProperties(propertiesToSetOrAdd, propertiesToRemove);
                 return this.Root;
             }
 
             try
             {
-                (_, CloudBlobContainer container) = await this.GetContainerAndTenantForChildTenantsOf(tenant.GetRequiredParentId()).ConfigureAwait(false);
+                (_, CloudBlobContainer container) = await this.GetContainerAndTenantForChildTenantsOf(TenantExtensions.GetRequiredParentId(tenantId)).ConfigureAwait(false);
+
+                CloudBlockBlob blob = GetLiveTenantBlockBlobReference(tenantId, container);
+
+                Tenant tenant = await this.GetTenantFromContainerAsync(tenantId, container, null).ConfigureAwait(false);
+
+                IPropertyBag updatedProperties = this.propertyBagFactory.CreateModified(
+                    tenant.Properties,
+                    propertiesToSetOrAdd.Select(kv => new KeyValuePair<string, object?>(kv.Key, kv.Value)),
+                    propertiesToRemove);
 
-                CloudBlockBlob blob = GetLiveTenantBlockBlobReference(tenant.Id, container);
-                string text = JsonConvert.SerializeObject(tenant, this.serializerSettings);
+                var updatedTenant = new Tenant(tenant.Id, tenant.Name, updatedProperties);
```

Comment:
So, there seems to be no point in the ITenant abstraction any more - we are explicitly creating Tenant instances, rather than just deserializing the (potentially polymorphic) type that was stored.

Reply:
`RootTenant` pops up from time to time, although that derives from `Tenant`. `Marain.Tenancy` defines its own custom `ITenant` that is uses to redact settings it has attached to the root tenant. (And that doesn't derive from `Tenant`. So there's at least one extant implementation of `ITenant` that isn't the `Tenant` type.)

It's true the the `TenantProviderBlobStore` constructs tenants directly. But you'd expect that because it's one of the two reusable things we currently have that acts as a source of tenants (the other being `Marain.Tenancy.Client`). It has to construct something. But the interfaces it implements, `ITenantProvider` and `ITenantStore` don't oblige it to supply any particular implementation.

That said, we certaily could get rid of `ITenant` at this point, because all the current use cases don't absolutely require the interface. `RootTenant` derives from `Tenant`. I think the redacted root tenant could in fact just be a plain `Tenant`. The question is whether we want to. I would argue it's useful for things like tenanted storage providers to be able to have a type they can use to represent a tenant without needing to care where it came from.

By re-expressing everything in terms of the concrete `Tenant`, the only path to bringing your own more interesting definition of tenant is to derive from that `Tenant` type. Which would work.

I just think I slightly prefer the idea of keeping the flexibility of an interface. I could imagine an entirely different implementation: something that's an of `Memory<T>` reference into a buffer (e.g., of JSON in UTF8 form) kept in a pool with the ability to change which bit of memory they point to as they go into and back out of the pool, so you could have an `ITenant` the does zero-allocations unless you actually ask it to do something. `ITenant` requires that you have actual `string` objects with the real values in whether anything needs them or not.